File: frontend/src/lib/gstr2bAutoMatch.ts
import type { MatchConfidence } from "./columnAutoMatch";
import { isGstr2bMappingSet, isGstr2bNaMapping } from "./gstr2bTabs";

const GSTR2B_FIELD_PATTERNS: Record<string, string[]> = {
  company_gstin: ["company gstin", "gstin of recipient", "recipient gstin", "gstin"],
  supplier_name: ["trade/legal name", "trade legal name", "legal name", "supplier name", "name of supplier"],
  supplier_gstin: ["gstin of supplier", "supplier gstin", "gstin"],
  supplier_invoice_no: ["invoice number", "invoice no", "note number", "document number", "document no", "invoice #", "number"],
  supplier_invoice_date: ["invoice date", "note date", "document date", "date"],
  taxable_amount: ["taxable value (₹)", "taxable value", "taxable amount", "taxable value(rs)"],
  igst: ["integrated tax(₹)", "integrated tax", "igst", "integrated tax amount"],
  sgst: ["state/ut tax(₹)", "state/ut tax", "state tax", "sgst", "state/ut tax amount"],
  cgst: ["central tax(₹)", "central tax", "cgst", "central tax amount"],
  total_tax: ["total tax", "tax amount", "total gst", "total tax(₹)"],
  grand_total: ["invoice value(₹)", "invoice value", "note value (₹)", "note value", "document value(₹)", "document value", "grand total", "total invoice value"],
};

const AMENDMENT_TABS = new Set(["B2BA", "B2B-CDNRA", "ECOA", "ISDA", "IMPGA", "IMPSEZA"]);

const NOTE_TYPE_PATTERNS = ["note type", "credit/debit note type", "credit debit note type", "type of note"];

function normalizeHeader(value: string): string {
  return value.trim().toLowerCase().replace(/[^a-z0-9]/g, "");
}

function patternScore(normalizedHeader: string, pattern: string): number {
  const normPattern = normalizeHeader(pattern);
  if (!normPattern) return 0;
  if (normalizedHeader === normPattern) return 100;
  if (normalizedHeader.endsWith(normPattern) || normalizedHeader.startsWith(normPattern)) return 80;
  if (normPattern.includes(normalizedHeader) || normalizedHeader.includes(normPattern)) return 60;
  return 0;
}

export function autoMatchGstr2bColumns(
  fieldCodes: string[],
  excelColumns: string[],
  includeNoteType = false,
  tab?: string,
): { mappings: Record<string, string | null>; confidence: Record<string, MatchConfidence> } {
  const usedColumns = new Set<string>();
  const mappings: Record<string, string | null> = {};
  const confidence: Record<string, MatchConfidence> = {};
  const normalizedColumns = Object.fromEntries(excelColumns.map((col) => [col, normalizeHeader(col)]));

  for (const fieldCode of fieldCodes) {
    const patterns = GSTR2B_FIELD_PATTERNS[fieldCode] ?? [];
    let bestColumn: string | null = null;
    let bestScore = 0;
    for (const col of excelColumns) {
      if (usedColumns.has(col)) continue;
      const normCol = normalizedColumns[col];
      for (const pattern of patterns) {
        const score = patternScore(normCol, pattern);
        if (score > bestScore) {
          bestScore = score;
          bestColumn = col;
        }
      }
    }
    if (bestColumn && bestScore >= 80) {
      let chosen = bestColumn;
      if (tab && AMENDMENT_TABS.has(tab)) {
        const base = chosen.replace(" (2)", "").trim();
        const revised = excelColumns.find((c) => c.startsWith(base) && c.includes("(2)"));
        if (
          revised &&
          ["supplier_invoice_no", "supplier_invoice_date", "igst", "cgst", "sgst", "grand_total"].includes(fieldCode)
        ) {
          chosen = revised;
        }
      }
      mappings[fieldCode] = chosen;
      confidence[fieldCode] = "high";
      usedColumns.add(chosen);
    } else if (bestColumn && bestScore >= 60) {
      mappings[fieldCode] = bestColumn;
      confidence[fieldCode] = "low";
      usedColumns.add(bestColumn);
    } else {
      mappings[fieldCode] = null;
      confidence[fieldCode] = "none";
    }
  }

  if (includeNoteType) {
    let bestColumn: string | null = null;
    let bestScore = 0;
    for (const col of excelColumns) {
      if (usedColumns.has(col)) continue;
      const normCol = normalizedColumns[col];
      for (const pattern of NOTE_TYPE_PATTERNS) {
        const score = patternScore(normCol, pattern);
        if (score > bestScore) {
          bestScore = score;
          bestColumn = col;
        }
      }
    }
    if (bestColumn && bestScore >= 60) {
      mappings.note_type = bestColumn;
      confidence.note_type = bestScore >= 80 ? "high" : "low";
    } else {
      mappings.note_type = null;
      confidence.note_type = "none";
    }
  }

  return { mappings, confidence };
}

export function isTotalTaxDerivable(mappings: Record<string, string | null>): boolean {
  const mapped = (code: string) => isGstr2bMappingSet(mappings[code]) && !isGstr2bNaMapping(mappings[code]);
  return mapped("igst") && mapped("cgst") && mapped("sgst");
}

export function isFieldRequiredForGstr2b(
  fieldCode: string,
  isRequired: boolean,
  mappings: Record<string, string | null>,
): boolean {
  if (!isRequired) return false;
  if (isGstr2bNaMapping(mappings[fieldCode])) return false;
  if (fieldCode === "company_gstin") return false;
  if (fieldCode === "total_tax") {
    if (isGstr2bMappingSet(mappings.total_tax) && !isGstr2bNaMapping(mappings.total_tax)) return false;
    return !isTotalTaxDerivable(mappings);
  }
  return !isGstr2bMappingSet(mappings[fieldCode]);
}

export function findDuplicateColumnMappingsForTab(
  mappings: Record<string, string | null>,
): string | null {
  const seen = new Map<string, string>();
  for (const [fieldCode, col] of Object.entries(mappings)) {
    if (!col || !isGstr2bMappingSet(col) || isGstr2bNaMapping(col)) continue;
    if (seen.has(col)) {
      return `Column "${col}" is mapped to both ${seen.get(col)} and ${fieldCode}.`;
    }
    seen.set(col, fieldCode);
  }
  return null;
}

File: backend/app/services/gstr2b_auto_match.py
"""Auto-match GSTR-2B GST portal column headers to master field codes."""

from __future__ import annotations

import re

from app.services.column_auto_match import Confidence, _normalize_header, _pattern_score

GSTR2B_FIELD_PATTERNS: dict[str, tuple[str, ...]] = {
    "company_gstin": (
        "company gstin",
        "gstin of recipient",
        "recipient gstin",
        "gstin",
    ),
    "supplier_name": (
        "trade/legal name",
        "trade legal name",
        "legal name",
        "supplier name",
        "name of supplier",
    ),
    "supplier_gstin": (
        "gstin of supplier",
        "supplier gstin",
        "gstin",
    ),
    "supplier_invoice_no": (
        "invoice number",
        "invoice no",
        "note number",
        "document number",
        "document no",
        "invoice #",
        "number",
    ),
    "supplier_invoice_date": (
        "invoice date",
        "note date",
        "document date",
        "date",
    ),
    "taxable_amount": (
        "taxable value (₹)",
        "taxable value",
        "taxable amount",
        "taxable value(rs)",
    ),
    "igst": (
        "integrated tax(₹)",
        "integrated tax",
        "igst",
        "integrated tax amount",
    ),
    "sgst": (
        "state/ut tax(₹)",
        "state/ut tax",
        "state tax",
        "sgst",
        "state/ut tax amount",
    ),
    "cgst": (
        "central tax(₹)",
        "central tax",
        "cgst",
        "central tax amount",
    ),
    "total_tax": (
        "total tax",
        "tax amount",
        "total gst",
        "total tax(₹)",
    ),
    "grand_total": (
        "invoice value(₹)",
        "invoice value",
        "note value (₹)",
        "note value",
        "document value(₹)",
        "document value",
        "grand total",
        "total invoice value",
    ),
}

AMENDMENT_TABS = frozenset({"B2BA", "B2B-CDNRA", "ECOA", "ISDA", "IMPGA", "IMPSEZA"})

NOTE_TYPE_PATTERNS: tuple[str, ...] = (
    "note type",
    "note type ",
    "credit/debit note type",
    "credit debit note type",
    "type of note",
)


def _match_single_field(
    patterns: tuple[str, ...],
    excel_columns: list[str],
    used_columns: set[str],
) -> tuple[str | None, Confidence, int]:
    normalized_columns = {col: _normalize_header(col) for col in excel_columns}
    best_column: str | None = None
    best_score = 0
    for col, norm_col in normalized_columns.items():
        if col in used_columns:
            continue
        for pattern in patterns:
            score = _pattern_score(norm_col, pattern)
            if score > best_score:
                best_score = score
                best_column = col
    if best_column and best_score >= 80:
        return best_column, "high", best_score
    if best_column and best_score >= 60:
        return best_column, "low", best_score
    return None, "none", 0


def _prefer_amendment_column(
    field_code: str,
    matched: str | None,
    excel_columns: list[str],
    tab: str | None,
) -> str | None:
    if not matched or not tab or tab not in AMENDMENT_TABS:
        return matched
    if field_code not in {
        "supplier_invoice_no",
        "supplier_invoice_date",
        "igst",
        "cgst",
        "sgst",
        "grand_total",
        "note_type",
    }:
        return matched
    base = matched.replace(" (2)", "").strip()
    for col in excel_columns:
        if col.startswith(base) and "(2)" in col:
            return col
    return matched


def auto_match_gstr2b_columns(
    field_codes: list[str],
    excel_columns: list[str],
    *,
    include_note_type: bool = False,
    tab: str | None = None,
) -> tuple[dict[str, str | None], dict[str, Confidence]]:
    used_columns: set[str] = set()
    mappings: dict[str, str | None] = {}
    confidence: dict[str, Confidence] = {}

    for field_code in field_codes:
        patterns = GSTR2B_FIELD_PATTERNS.get(field_code, ())
        col, conf, _ = _match_single_field(patterns, excel_columns, used_columns)
        col = _prefer_amendment_column(field_code, col, excel_columns, tab)
        mappings[field_code] = col
        confidence[field_code] = conf
        if col:
            used_columns.add(col)

    if include_note_type:
        col, conf, _ = _match_single_field(NOTE_TYPE_PATTERNS, excel_columns, used_columns)
        col = _prefer_amendment_column("note_type", col, excel_columns, tab)
        mappings["note_type"] = col
        confidence["note_type"] = conf
        if col:
            used_columns.add(col)

    return mappings, confidence
